### PR TITLE
New registry dependent interlayer potential in LAMMPS

### PR DESCRIPTION
**Pull request by oywg11**
## Purpose
Given the large interest of the scientific community in understanding the properties of these class of layered materials, I implemented a new registry-dependent interlayer potential (RDILP) into LAMMPS. This potential was initially proposed by Kolmogorov and Crepsi (Kolmogorov, A. N.; Crespi, V. H. Phys. Rev. B 2005, 71, 235415.) for graphitic systems and recently it has been generalized by Prof. Hod's group (J. Chem.Theory Comput. 2016, 12, 2896-905 and J. Phys. Chem. C 2017, 121, 22826-22835.) for both graphitic and boron nitride systems. The advantage of this potential is that it can give both a good description of the binding energy and of the potential energy surface under sliding. 

Here, I provide two versions of RDILP, the first one is with the form of Kolmogorov and Crepsi (KC). Note that although there is a version of this potential in LAMMPS (paie_kolmogorov_crepsi_z), it made a significant simplification that assumes all normals along the z-axis. The potential provided here removed this limitation, i.e., calculate the normals correctly even for curved surfaces. Besides KC potential, we also provide the RDILP with the form of Prof. Hod's group. This potential can be used for both graphitic and boron nitride systems. Again, no simplification is made.

Moreover, I improved the fitting procedure to the reference DFT data and defined a new set of parameters that provides a good description of interlayer interactions even in the high pressure (small interlayer distances) regime, previously not included. Currently, we are in the process of writing a paper presenting the new parameters, we will release the new parameters when the paper is published.

## Author(s)

Wengen Ouyang, School of Chemistry, Tel Aviv University.

## Backward Compatibility

The changes in the pull request do not break backward compatibility for inputs.

## Implementation Notes

The most difficult part is how to calculate the normals correctly. The method to calculate the normals in the present codes is averaging the three normalized cross products of the displacement vectors to the nearest neighbours. To do this, I build a neighbour list from the main neighbour list to define the nearest neighbours. After calculating the normals and its derivatives for local atoms, I use the functions pack_forward_comm and unpack_forward_comm to communicate the normals and its derivatives between local atoms and ghost atoms, this ensures they are calculated correctly. For the case of edge atoms, who only have one or two nearest neighbours, I also calculate its normals correctly. Another thing should be noted is that there is an indirect interaction between nearest neighbour atoms due to the coupling through the normals, this is also correctly calculated. I add comments as many as I can to make the codes easy to read.

I have already worked on it for over half a year and performed careful benchmark tests. The details are as follows: I compared the binding energy curves and sliding potential surfaces between my implementation and DFT data as well as our home-made codes, the results of my implementation show good agreement with DFT data (as illustrated in the paper) and got the exact same results with our home-made codes. The training set are three periodic structures (graphene/graphene, graphene/h-BN and h-BN/h-BN) and 10 finite structures (Benzene dimer, Borazine dimer, B12N12H12 dimer, Coronene dimer, Benzene/Coronene, Borazine/B12N12H12, Benzene/Borazine, Benzene/ B12N12H12, Borazine/Coronene and Coronene/B12N12H12). After this step, I did NVE simulation for all the structures to check the energy conservation. I’m now positively confident that the implementation is correct and without errors. The benchmark tests will be published in my recent paper.
Other features in LAMMPS are unaffected.

## Post Submission Checklist

_Please check the fields below as they are completed_
- [Yes ] The feature or features in this pull request is complete
- [Yes] Suitable new documentation files and/or updates to the existing docs are included
- [ Yes] One or more example input decks are included
- [ Yes] The source code follows the LAMMPS formatting guidelines

## Further Information, Files, and Links

_Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)_

 I. Leven et al, J. Chem.Theory Comput. 12, 2896-905 (2016)
 T. Maaravi et al, J. Phys. Chem. C 121, 22826-22835 (2017).
 A. N. Kolmogorov, V. H. Crespi, Phys. Rev. B 71, 235415 (2005)

### CODE REVIEW

**Comment by akohlmey** [comment]
@oywg11 i've done some modifications to style names, corresponding file name examples and cleaned up the code a little bit and make it more in sync with other lammps styles.

**Comment by oywg11** [comment]
@akohlmey Thanks! I refined the code and documentation according to your modifications and added a potential file (CC.KC-full) for pair style kolmogorov/crespi/full.

**Comment by akohlmey** [comment]
@oywg11 thanks for the additional updates and the potential file. this is a very well done pull request and it looks like the code is working nicely, too. i also updated the reference runs using those additional parameters.

**Comment by oywg11** [comment]
@akohlmey Thanks for your comments and careful check! I will update two additional sets of parameters which give better performance when our manuscript is accepted.
